File: modules/agent/harness_checkpoint_file.go
package agent

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// FileCheckpointStore persists harness checkpoints as JSON files.
type FileCheckpointStore struct {
	dir string
	ttl time.Duration // Phase 14: optional TTL for automatic cleanup
}

// NewFileCheckpointStore creates a file-backed checkpoint store.
func NewFileCheckpointStore(dir string) *FileCheckpointStore {
	return &FileCheckpointStore{dir: dir}
}

// NewFileCheckpointStoreWithTTL creates a file-backed store with automatic TTL cleanup.
func NewFileCheckpointStoreWithTTL(dir string, ttl time.Duration) *FileCheckpointStore {
	return &FileCheckpointStore{dir: dir, ttl: ttl}
}

// Save writes a checkpoint atomically using write-to-temp + rename.
func (s *FileCheckpointStore) Save(ctx context.Context, cp HarnessCheckpoint) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	if cp.RunID == "" {
		return fmt.Errorf("file checkpoint store: run id is required")
	}
	if err := os.MkdirAll(s.dir, 0o755); err != nil {
		return fmt.Errorf("file checkpoint store: mkdir: %w", err)
	}
	cp = cloneHarnessCheckpoint(cp)
	cp.UpdatedAt = time.Now().UTC()
	data, err := json.MarshalIndent(cp, "", "  ")
	if err != nil {
		return fmt.Errorf("file checkpoint store: marshal: %w", err)
	}

	// Phase 14: Atomic write via temp file + rename
	tmpPath := s.pathFor(cp.RunID) + ".tmp"
	if err := os.WriteFile(tmpPath, data, 0o644); err != nil {
		return fmt.Errorf("file checkpoint store: write temp: %w", err)
	}
	if err := os.Rename(tmpPath, s.pathFor(cp.RunID)); err != nil {
		_ = os.Remove(tmpPath) // cleanup temp file
		return fmt.Errorf("file checkpoint store: rename: %w", err)
	}
	return nil
}

// Load reads a checkpoint by run ID.
func (s *FileCheckpointStore) Load(ctx context.Context, runID string) (HarnessCheckpoint, error) {
	if err := ctx.Err(); err != nil {
		return HarnessCheckpoint{}, err
	}
	data, err := os.ReadFile(s.pathFor(runID))
	if err != nil {
		return HarnessCheckpoint{}, fmt.Errorf("file checkpoint store: read: %w", err)
	}
	var cp HarnessCheckpoint
	if err := json.Unmarshal(data, &cp); err != nil {
		return HarnessCheckpoint{}, fmt.Errorf("file checkpoint store: unmarshal: %w", err)
	}
	return cloneHarnessCheckpoint(cp), nil
}

// List returns summaries for checkpoints matching the given prefix.
func (s *FileCheckpointStore) List(ctx context.Context, prefix string) ([]CheckpointSummary, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	entries, err := os.ReadDir(s.dir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("file checkpoint store: read dir: %w", err)
	}

	var summaries []CheckpointSummary
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".json") {
			continue
		}
		runID := strings.TrimSuffix(entry.Name(), ".json")
		if prefix != "" && !hasPrefix(runID, prefix) {
			continue
		}

		// Read only enough to build summary
		data, err := os.ReadFile(s.pathFor(runID))
		if err != nil {
			continue
		}
		var cp HarnessCheckpoint
		if err := json.Unmarshal(data, &cp); err != nil {
			continue
		}

		// TTL check
		if s.ttl > 0 && time.Since(cp.UpdatedAt) > s.ttl {
			_ = s.Delete(ctx, runID)
			continue
		}

		summaries = append(summaries, cp.Summary())
	}

	// Sort by UpdatedAt descending (most recent first)
	sort.Slice(summaries, func(i, j int) bool {
		return summaries[i].UpdatedAt.After(summaries[j].UpdatedAt)
	})
	return summaries, nil
}

// Delete removes a checkpoint file by run ID.
func (s *FileCheckpointStore) Delete(ctx context.Context, runID string) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	path := s.pathFor(runID)
	if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("file checkpoint store: delete: %w", err)
	}
	return nil
}

// CleanupExpired removes all checkpoints older than the configured TTL.
func (s *FileCheckpointStore) CleanupExpired(ctx context.Context) (int, error) {
	if s.ttl <= 0 {
		return 0, nil
	}
	entries, err := os.ReadDir(s.dir)
	if err != nil {
		if os.IsNotExist(err) {
			return 0, nil
		}
		return 0, fmt.Errorf("file checkpoint store: cleanup read dir: %w", err)
	}
	cleaned := 0
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".json") {
			continue
		}
		runID := strings.TrimSuffix(entry.Name(), ".json")
		data, err := os.ReadFile(s.pathFor(runID))
		if err != nil {
			continue
		}
		var cp HarnessCheckpoint
		if err := json.Unmarshal(data, &cp); err != nil {
			continue
		}
		if time.Since(cp.UpdatedAt) > s.ttl {
			if err := s.Delete(ctx, runID); err != nil {
				return cleaned, err
			}
			cleaned++
		}
	}
	return cleaned, nil
}

func (s *FileCheckpointStore) pathFor(runID string) string {
	return filepath.Join(s.dir, runID+".json")
}

File: packages/agent/src/harness-checkpoint-file.ts
/**
 * FileCheckpointStore persists harness checkpoints as JSON files.
 * Ported from modules/agent/harness_checkpoint_file.go.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import type { HarnessCheckpoint, CheckpointSummary } from "./harness-state.js";
import { checkpointSummary, cloneHarnessCheckpoint } from "./harness-state.js";
import type { CheckpointStore } from "./harness-state.js";

export class FileCheckpointStore implements CheckpointStore {
  private _dir: string;
  private _ttlMs: number; // 0 = no TTL

  constructor(dir: string, ttlMs?: number) {
    this._dir = dir;
    this._ttlMs = ttlMs ?? 0;
  }

  /** Save writes a checkpoint atomically using write-to-temp + rename. */
  async save(_signal: AbortSignal | undefined, cp: HarnessCheckpoint): Promise<void> {
    if (!cp.runID) throw new Error("file checkpoint store: run id is required");
    await fs.promises.mkdir(this._dir, { recursive: true });

    const cloned = cloneHarnessCheckpoint(cp);
    cloned.updatedAt = new Date();
    const data = JSON.stringify(cloned, null, 2);

    // Atomic write via temp file + rename
    const tmpPath = this.pathFor(cp.runID) + ".tmp";
    await fs.promises.writeFile(tmpPath, data, "utf-8");
    try {
      await fs.promises.rename(tmpPath, this.pathFor(cp.runID));
    } catch (err) {
      await fs.promises.unlink(tmpPath).catch(() => {});
      throw new Error(`file checkpoint store: rename: ${err instanceof Error ? err.message : String(err)}`);
    }
  }

  /** Load reads a checkpoint by run ID. */
  async load(_signal: AbortSignal | undefined, runID: string): Promise<HarnessCheckpoint> {
    try {
      const data = await fs.promises.readFile(this.pathFor(runID), "utf-8");
      const cp = JSON.parse(data) as HarnessCheckpoint;
      return cloneHarnessCheckpoint(cp);
    } catch (err) {
      throw new Error(`file checkpoint store: read: ${err instanceof Error ? err.message : String(err)}`);
    }
  }

  /** List returns summaries for checkpoints matching the given prefix. */
  async list(_signal: AbortSignal | undefined, prefix: string): Promise<CheckpointSummary[]> {
    let entries: fs.Dirent[];
    try {
      entries = await fs.promises.readdir(this._dir, { withFileTypes: true });
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return [];
      throw new Error(`file checkpoint store: read dir: ${err instanceof Error ? err.message : String(err)}`);
    }

    const summaries: CheckpointSummary[] = [];
    for (const entry of entries) {
      if (entry.isDirectory() || !entry.name.endsWith(".json")) continue;
      const runID = entry.name.slice(0, -5); // remove .json
      if (prefix && !runID.startsWith(prefix)) continue;

      try {
        const data = await fs.promises.readFile(this.pathFor(runID), "utf-8");
        const cp = JSON.parse(data) as HarnessCheckpoint;

        // TTL check
        if (this._ttlMs > 0 && Date.now() - cp.updatedAt.getTime() > this._ttlMs) {
          await this.delete(undefined, runID).catch(() => {});
          continue;
        }

        summaries.push(checkpointSummary(cp));
      } catch {
        continue;
      }
    }

    // Sort by UpdatedAt descending (most recent first)
    summaries.sort((a, b) => b.updatedAt.getTime() - a.updatedAt.getTime());
    return summaries;
  }

  /** Delete removes a checkpoint file by run ID. */
  async delete(_signal: AbortSignal | undefined, runID: string): Promise<void> {
    const filePath = this.pathFor(runID);
    try {
      await fs.promises.unlink(filePath);
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "ENOENT") {
        throw new Error(`file checkpoint store: delete: ${err instanceof Error ? err.message : String(err)}`);
      }
    }
  }

  /** CleanupExpired removes all checkpoints older than the configured TTL. */
  async cleanupExpired(signal: AbortSignal | undefined): Promise<number> {
    if (this._ttlMs <= 0) return 0;
    let entries: fs.Dirent[];
    try {
      entries = await fs.promises.readdir(this._dir, { withFileTypes: true });
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return 0;
      throw err;
    }
    let cleaned = 0;
    for (const entry of entries) {
      if (entry.isDirectory() || !entry.name.endsWith(".json")) continue;
      const runID = entry.name.slice(0, -5);
      try {
        const data = await fs.promises.readFile(this.pathFor(runID), "utf-8");
        const cp = JSON.parse(data) as HarnessCheckpoint;
        if (Date.now() - cp.updatedAt.getTime() > this._ttlMs) {
          await this.delete(signal, runID);
          cleaned++;
        }
      } catch {
        continue;
      }
    }
    return cleaned;
  }

  private pathFor(runID: string): string {
    return path.join(this._dir, `${runID}.json`);
  }
}
